internal/controller/thanosquery_controller.go

```diff
@@ -132,29 +143,122 @@ func (r *ThanosQueryReconciler) buildQuerier(query monitoringthanosiov1alpha1.Th
 		LogFormat: query.Spec.LogFormat,
 	}.ApplyDefaults()
 
+	endpoints, err := r.getStoreAPIServiceEndpoints(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+
 	return manifestquery.BuildQuerier(manifestquery.QuerierOptions{
 		Options:       metaOpts,
 		ReplicaLabels: query.Spec.QuerierReplicaLabels,
 		Timeout:       "15m",
 		LookbackDelta: "5m",
 		MaxConcurrent: 20,
+		Endpoints:     endpoints,
+	}), nil
+}
 
-		// TODO(saswatamcode): Add logic to for automated tracking of endpoints.
-		Endpoints: []manifestquery.Endpoint{{
-			ServiceName: "thanos-receive",
-			Namespace:   query.GetNamespace(),
-			Type:        manifestquery.Regular,
-		}},
-	})
+// getStoreAPIServiceEndpoints returns the list of endpoints for the StoreAPI services that match the ThanosQuery storeLabelSelector.
+func (r *ThanosQueryReconciler) getStoreAPIServiceEndpoints(ctx context.Context, query monitoringthanosiov1alpha1.ThanosQuery) ([]manifestquery.Endpoint, error) {
+	selector, err := metav1.LabelSelectorAsSelector(query.Spec.StoreLabelSelector)
+	if err != nil {
+		return []manifestquery.Endpoint{}, err
+	}
+
+	services := &corev1.ServiceList{}
+	listOpts := client.ListOptions{
+		Namespace:     query.Namespace,
+		LabelSelector: selector,
+	}
+	if err := r.List(ctx, services, &listOpts); err != nil {
+		return []manifestquery.Endpoint{}, err
+	}
+
+	endpoints := make([]manifestquery.Endpoint, 0, len(services.Items))
+	for i, svc := range services.Items {
+		etype := manifestquery.Regular
+
+		if metav1.HasLabel(svc.ObjectMeta, string(manifestquery.Strict)) {
+			etype = manifestquery.Strict
+		} else if metav1.HasLabel(svc.ObjectMeta, string(manifestquery.GroupStrict)) {
+			etype = manifestquery.GroupStrict
+		} else if metav1.HasLabel(svc.ObjectMeta, string(manifestquery.Group)) {
+			etype = manifestquery.Group
+		}
+
+		metav1.LabelSelectorAsSelector(query.Spec.StoreLabelSelector)
+		endpoints[i] = manifestquery.Endpoint{
+			ServiceName: svc.GetName(),
+			Namespace:   svc.GetNamespace(),
+			Type:        etype,
+		}
+	}
+
+	return endpoints, nil
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ThanosQueryReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	servicePredicate, err := predicate.LabelSelectorPredicate(metav1.LabelSelector{
+		MatchLabels: map[string]string{manifests.PartOfLabel: manifests.DefaultPartOfLabel},
```

Comment:
couldnt we limit the further to the label added in https://github.com/thanos-community/thanos-operator/pull/16/files#diff-bc456696a313dcb9dc6239892cf53a51abf95de69c84136de0649cedaf4138fdR19 ?